### PR TITLE
Add type hints (Part 3)

### PR DESCRIPTION
**Pull request by desilinguist**
### Changes

- Add `statsmodels` intersphinx mapping for documentation.
- add type hints to `fairness_utils.py`.
- add type hints to `modeler.py`.
- add type hints to `preprocessor.py`.
- add type hints to `reader.py`
- Remove incorrect exceptions in some tests based on changes in `reader.py`. 

Note that in this MR, I also did some refactoring:
- Removed unused `id_column` argument in `FeaturePreprocessor.filter_on_column()`. 
- Changed the behavior of `DataReader.locate_files()` to always return list of string instead of either a string or a list of strings + `None`s which was confusing and also caused issues when adding type hints. 
- I also had to change a couple of tests to check for the raising of the correct exception type based on the changes in `DataReader`.

### How to review
- Please check that I have not left any parameters/arguments/return values without type hints.
- Make sure that I didn't screw up anything when doing the refactoring. 
- Here's the [RTD build for this branch](https://rsmtool.readthedocs.io/en/326-type-hints-3). Check the "API Documentation" section (only for the files affected by this MR) to confirm that all types are linked to relevant pages from external documentations (e.g., Python, Numpy, SKLL, stats models etc.)
- Check for typos and/or weird issues in docstrings.

### CODE REVIEW

**Comment by desilinguist** [comment]
We can ignore the coverage failure.

File: rsmtool/configuration_parser.py [hunk 1]
```diff
@@ -560,9 +560,9 @@ def get_trim_min_max_tolerance(self) -> Tuple[float, float, float]:
 
         Returns
         -------
-        spec_trim_min : float
+        spec_trim_min : int
```
[the remark this answers is not in the document]
Comment (tamarl08) [reply rsmtool/configuration_parser.py:563]: oh and it should be Optional[int], right?

**Review by mulhod — APPROVED** [review]

File: rsmtool/modeler.py [hunk 2]
```diff
@@ -34,25 +37,27 @@
...
 
-    def __init__(self, logger=None):
+    def __init__(self, logger: Optional[logging.Logger] = None) -> None:
         """
-        Initialize empty ``Modeler`` object.
+        Initialize empty Modeler object.
 
-        The learner and the scaling-/trimming-related attributes are
-        set to ``None``.
+        The learner and scaling-/trimming-related attributes are set to ``None``.
```
Comment (tamarl08) [inline rsmtool/modeler.py:44]: missing `Parameters`

File: rsmtool/preprocessor.py [hunk 3]
```diff
@@ -287,8 +294,13 @@ def validate_feature_specs(self, df, use_truncations=False):
         return df_specs_new
 
     def generate_specs(
-        self, df, feature_names, train_label, feature_subset=None, feature_sign=None
-    ):
+        self,
+        df: pd.DataFrame,
+        feature_names: List[str],
+        train_label: str,
+        feature_subset: pd.DataFrame = None,
```
Comment (tamarl08) [inline rsmtool/preprocessor.py:301]:
```suggestion
        feature_subset: Optional[pd.DataFrame] = None,
```

File: rsmtool/preprocessor.py [hunk 4]
```diff
@@ -390,27 +408,35 @@ def trim(self, values, trim_min, trim_max, tolerance=0.4998):
...
+
+        Raises
+        ------
+        ValueError
+            If ``trim_min``, ``trim_max``, or ``tolerance`` are not numeric.
         """
         if isinstance(values, list):
             values = np.array(values)
 
+        if trim_min is None or trim_max is None or tolerance is None:
```
Comment (tamarl08) [inline rsmtool/preprocessor.py:437]: is None the right condition to test here?  these params are listed as optional.

File: rsmtool/preprocessor.py [hunk 5]
```diff
@@ -479,18 +508,17 @@ def remove_outliers_using_truncations(self, values, feature_name, truncations):
```
Comment (tamarl08) [inline rsmtool/preprocessor.py:508]: Missing some info in the docstring: IIUC another difference from `remove_outliers` is that it processes just one feature? 
and it is assumed that the given data is only values of `feature_name`?

File: rsmtool/preprocessor.py [hunk 6]
```diff
@@ -1297,70 +1334,70 @@ def filter_data(
...
             The candidate column in the data.
-        requested_feature_names : list
+        requested_feature_names : List[str]
             A list of requested feature names.
-        reserved_column_names : list
+        reserved_column_names : List[str]
             A list of reserved column names.
         given_trim_min : float
             The minimum trim value.
         given_trim_max : float
```
Comment (tamarl08) [inline rsmtool/preprocessor.py:1355]:
```suggestion
        given_trim_max : Optional[float]
```

File: rsmtool/reader.py [hunk 7]
```diff
@@ -112,31 +119,37 @@ def try_to_load_file(filename, converters=None, raise_error=False, raise_warning
     FileNotFoundError
         If ``raise_error`` is ``True`` and the file cannot be located.
     """
-    if exists(filename):
-        return DataReader.read_from_file(filename, converters, **kwargs)
-
     message = f"The file '{filename}' could not be located."
```
Comment (tamarl08) [inline rsmtool/reader.py:122]: define message inside the `if`

File: rsmtool/reader.py [hunk 8]
```diff
@@ -112,31 +119,37 @@ def try_to_load_file(filename, converters=None, raise_error=False, raise_warning
...
+    ):
         """
         Initialize a DataReader object.
 
         Parameters
         ----------
-        filepaths : list of str
-            A list of paths to files that are to be read in.
-        framenames : list of str
+        filepaths : List[Optional[str]]
```
Comment (tamarl08) [inline rsmtool/reader.py:147]: it is not optional above. which is right?

File: rsmtool/modeler.py [hunk 9]
```diff
@@ -1331,14 +1367,14 @@ def predict_train_and_test(self, df_train, df_test, configuration):
 
         df_train_predictions = self.predict(
             df_train,
-            min_score=int(trim_min),
-            max_score=int(trim_max),
+            min_score=trim_min,
```
[the remark this answers is not in the document]
Comment (desilinguist) [reply rsmtool/modeler.py:1370]: No, everything is passed as floats all the way except in one function where we need them to be ints, so we don't need to cast them here.

**Review by tamarl08 — APPROVED** [review]

**Pull request merged**